nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -842,25 +890,77 @@ ngx_http_modsecurity_detection_task_offload(ngx_http_request_t *r)
 }
 
 static void
-ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
+ngx_http_modsecurity_body_handler(ngx_http_request_t* r)
 {
     r->main->count--;
     ngx_http_core_run_phases(r);
 }
 
+static ngx_int_t
+ngx_http_calculate_modsec_latency(ngx_http_request_t *r, ngx_time_t *start_time,
+                                  ngx_http_modsecurity_loc_conf_t * cf)
+{
+    ngx_time_t *end_time;
+    ngx_msec_int_t   ms;
+    ngx_int_t ret;
+    ngx_time_update();
+    end_time = ngx_timeofday();
+
+    ms = (ngx_msec_int_t)
+        ((end_time->sec - start_time->sec) * 1000 + (end_time->msec - start_time->msec));
+    ms = ngx_max(ms, 0);
+
+    ngx_http_variable_value_t* modsec_latency_detect_var = ngx_http_get_indexed_variable(r, modsec_latency_detect_index);
+    ngx_http_variable_value_t* modsec_latency_prevent_var = ngx_http_get_indexed_variable(r, modsec_latency_prevent_index);
+    if (cf->config->is_enabled == MODSEC_DETECTION_ONLY) {
+        ret = ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_detect_var, ms);
+        if (ret != NGX_OK) {
+            return ret;
+        }
+        return ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_prevent_var, 0);
+    }
+    else {
+        ret = ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_prevent_var, ms);
+        if (ret != NGX_OK) {
+            return ret;
+        }
+        return ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_detect_var, 0);
+    }
+}
 
 /*
 ** [ENTRY POINT] does : this function called by nginx from the request handler
+*   Calculate the modsec latency in this function and invoke the main handler.
 */
 static ngx_int_t
-ngx_http_modsecurity_handler(ngx_http_request_t *r)
+ngx_http_modsecurity_handler_with_timer(ngx_http_request_t* r)
 {
-    ngx_http_modsecurity_loc_conf_t *cf;
-    ngx_http_modsecurity_ctx_t      *ctx;
-    ngx_int_t                        rc;
+    ngx_time_update();
+    ngx_http_modsecurity_loc_conf_t* cf;
+    ngx_time_t* start_time;
+    ngx_int_t ret, modsec_latency_ret;
+    start_time = ngx_timeofday();
 
     cf = ngx_http_get_module_loc_conf(r, ngx_http_modsecurity);
 
+    // Main modsec handler
+    ret  = ngx_http_modsecurity_handler(r, cf);
+
+    // We return failure only if memory allocation fails for latency variable in calculating metric
+    modsec_latency_ret = ngx_http_calculate_modsec_latency(r, start_time, cf);
+    if (modsec_latency_ret != NGX_OK) {
+        ngx_log_error(NGX_LOG_INFO, r->connection->log, 0, "modSecurity: latency metric memory allocation failed");
+        return modsec_latency_ret;
```

Comment:
This should not be here.
For prevention... 
We would call 

ngx_http_modsecurity_handler ->ngx_http_modsecurity_prevention_task_offload ->....->ngx_http_modsecurity_prevention_thread_func->..ngx_http_modsecurity_prevention_thread_completion ->ngx_http_modsecurity_handler 

We would return 
`    if (ctx->status_code != STATUS_CODE_NOT_SET) {
        if (ctx->status_code > 0) {
            //calculate end time and return
            return ctx->status_code;
        }

        // Request must be routed to the next handler
        return NGX_DECLINED;
    }`

Ensure we do not override the start time with every call to ngx_http_modsecurity_handler .
Also as far as I remember we will call ngx_http_modsecurity_handler  again for body :)

Reply:
Moved the start time inside the modsec request ctx so that it remains consistent across multiple calls to the ngx_http_modsecurity_handler.
We calculate latency after every request completion and overwrite the variable.
The last write will be the final result and the max latency and that is what we are interested in.
Since there are multiple paths, this was the cleanest way to keep it neat. Otherwise we have to patch multiple paths with latency calculations.

Reply:
I like this approach @karanbazaz . It looks much cleaner. We just have to ensure its correctness.